src/kernel/schedule.c

```diff
@@ -154,13 +180,58 @@ NOTRACE void __attribute__((noreturn)) runloop_internal()
 
         /* should be a list of per-runloop checks - also low-pri background */
         mm_service();
-        update_timer();
+        timer_updated = update_timer();
 
         kern_unlock();
     }
 
-    if (!shutting_down && (t = dequeue(thread_queue)) != INVALID_ADDRESS)
-        run_thunk(t, cpu_user);
+    if (!shutting_down) {
+        t = dequeue(ci->thread_queue);
+        if (t == INVALID_ADDRESS) {
+            if (idle_cpu_mask) {
+                /* Try to steal a thread from an idle CPU (so that it doesn't
+                 * have to be woken up), and wake up CPUs that have a non-empty
+                 * thread queue). */
+                t = migrate_to_self(t, idle_cpu_mask & ~MASK(ci->id + 1));
+                t = migrate_to_self(t, idle_cpu_mask & MASK(ci->id));
+            }
+            if (t == INVALID_ADDRESS) {
+                /* No threads found in idle CPUs: try to steal a thread from a
+                 * CPU that is currently running another thread. */
+                for (u64 cpu = ci->id + 1; ; cpu++) {
+                    if (cpu == total_processors)
+                        cpu = 0;
+                    if (cpu == ci->id)
+                        break;
+                    cpuinfo cpui = cpuinfo_from_id(cpu);
+                    if (cpui->state == cpu_user) {
+                        t = dequeue(cpui->thread_queue);
+                        if (t != INVALID_ADDRESS) {
+                            sched_debug("migrating thread from CPU %d to self\n", cpu);
+                            break;
+                        }
+                    }
+                }
+            }
+        } else if (idle_cpu_mask) {
+            /* Wake up idle CPUs that have a non-empty thread queue, and if our
+             * thread queue is non-empty, migrate our threads to idle CPUs. */
+            migrate_from_self(ci, idle_cpu_mask & ~MASK(ci->id + 1));
+            migrate_from_self(ci, idle_cpu_mask & MASK(ci->id));
+        }
+        if (t != INVALID_ADDRESS) {
+            if (!timer_updated && (total_processors > 1)) {
+                timestamp here = now(CLOCK_ID_MONOTONIC);
+                s64 timeout = ci->last_timer_update - here;
+                if ((timeout < 0) || (timeout > runloop_timer_max)) {
```

Comment:
Just to verify: so timeout < 0 means that the last timer update on this cpu already passed, but a scheduler timer was not set - so set one now. Is that right?

Reply:
Yes, correct.